### PR TITLE
Nuke boost::thread and boost::thread_group

### PR DESCRIPTION
**Pull request by theuni**
This replaces #8023 with a solution that moves us directly to std::thread, rather than using an intermediate wrapper.
### Background

std::thread is pretty much a drop-in replacement for boost::thread, except that boost::thread is interruptible. We've used those interruptions, but we'll have to drop them eventually as they didn't become part of the spec.

When an interruption point is hit, it throws an instance of boost::thread_interrupted. boost::thread catches these automatically by default.

The interruption points are defined as:

```

    boost::thread::join()
    boost::thread::timed_join()
    boost::thread::try_join_for(),
    boost::thread::try_join_until(),
    boost::condition_variable::wait()
    boost::condition_variable::timed_wait()
    boost::condition_variable::wait_for()
    boost::condition_variable::wait_until()
    boost::condition_variable_any::wait()
    boost::condition_variable_any::timed_wait()
    boost::condition_variable_any::wait_for()
    boost::condition_variable_any::wait_until()
    boost::thread::sleep()
    boost::this_thread::sleep_for()
    boost::this_thread::sleep_until()
    boost::this_thread::interruption_point()
```

The ones relevant to us are primarily `boost::condition_variable_any::wait()`, `boost::this_thread::sleep_for()`, and of course `boost::this_thread::interruption_point()`

Any boost::thread will immediately throw boost::thread_interrupted if those are hit at any point. So we can't simply s/boost::thread/std::thread/g, as we would never be able to exit.
### Changes
- All threads are now externally interruptible. Rather than having a single bool for shutting down, I've attempted to notify each relevant subsystem. But because they're not well-defined, the notifications are somewhat arbitrary as well.
- For long-running threads, our own `thread_interrupted` and `interruption_point` have been added.
- Threads should call TraceThread if they may hit an interruption_point.
- MilliSleep is no longer interruptible.
- boost::thread_group is gone. All threads are now managed individually
### Future changes

There is lots of cleanup to be done post-merge. I was very tempted to include cleanups here, but I've elected to keep the changes minimal. Todo:
- s/boost::mutex/std::mutex/
- s/boost::condition_variable/std::condition_variable/
- s/boost::chrono/std::chrono/
- Drop a ton of boost includes
- Drop other boost cruft (like catching `boost::thread_interrupted`)

Also, I started on this by creating a base class for launching threads. Each instance would be responsible for implementing Interrupt() and Stop(). But because this is already a complex set of changes, I think it's best to do that as a separate step.

### CODE REVIEW

**Comment by JeremyRubin** [comment]
Concept ACK. 

Will review & test soon.

The boost interruption_point paradigm is messy and can make code hard to reason about. In my code for https://github.com/bitcoin/bitcoin/pull/8464 I have also removed boost::thread and interruption points, preferring more explicit quitting. Dealing with the interruption code made working on #8464  more difficult than it had to be.

One Nit, which I think is worth addressing. I would refactor the names from Interrupt to Quit, Shutdown, or even ControlledExit. Connotation of interrupt is a non-consensual break, whereas Quit or Shutdown implies being done as a normal part of control flow.

**Comment by theuni** [comment]
@JeremyRubin Nearly every thread uses separate interrupt/shutdown calls for exactly that reason.

Interrupt is non-consensual, and means "stop what you're doing and don't start anything else". Shutdown means "please gracefully exit now and don't return until you do".

**Comment by JeremyRubin** [comment]
@theuni Ah, gotcha. Was looking at small sample; my bad. I guess my comment should instead be, we should eventually migrate to shutdown only no interrupt threads; but this is a big step in that direction and precursor to that happening.

**Comment by sipa** [comment]
Concept ACK, testing.

**Comment by paveljanik** [comment]
OS X here:

```
Making all in src
  CXX      test/test_test_bitcoin-scheduler_tests.o
In file included from test/scheduler_tests.cpp:8:
In file included from ./test/test_bitcoin.h:16:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/thread:337:5: error: attempt to use a deleted function
    __invoke(std::__1::move(std::__1::get<0>(__t)), std::__1::move(std::__1::get<_Indices>(__t))...);
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/thread:347:5: note: in instantiation of function template specialization 'std::__1::__thread_execute<void (CScheduler::*)(), std::__1::reference_wrapper<CScheduler> , 1>' requested here
    __thread_execute(*__p, _Index());
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/thread:359:42: note: in instantiation of function template specialization 'std::__1::__thread_proxy<std::__1::tuple<void (CScheduler::*)(), std::__1::reference_wrapper<CScheduler> > >' requested here
    int __ec = pthread_create(&__t_, 0, &__thread_proxy<_Gp>, __p.get());
                                         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:1673:31: note: in instantiation of function template specialization 'std::__1::thread::thread<void (CScheduler::*)(), std::__1::reference_wrapper<CScheduler> , void>' requested here
            ::new((void*)__p) _Up(std::__1::forward<_Args>(__args)...);
                              ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:1600:18: note: in instantiation of function template specialization 'std::__1::allocator<std::__1::thread>::construct<std::__1::thread, void (CScheduler::*)(), std::__1::reference_wrapper<CScheduler> >' requested here
            {__a.construct(__p, std::__1::forward<_Args>(__args)...);}
                 ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:1453:14: note: in instantiation of function template specialization 'std::__1::allocator_traits<std::__1::allocator<std::__1::thread> >::__construct<std::__1::thread, void (CScheduler::*)(), std::__1::reference_wrapper<CScheduler> >' requested here
            {__construct(__has_construct<allocator_type, _Tp*, _Args...>(),
             ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/vector:1643:25: note: in instantiation of function template specialization 'std::__1::allocator_traits<std::__1::allocator<std::__1::thread> >::construct<std::__1::thread, void (CScheduler::*)(), std::__1::reference_wrapper<CScheduler> >' requested here
        __alloc_traits::construct(this->__alloc(),
                        ^
test/scheduler_tests.cpp:89:22: note: in instantiation of function template specialization 'std::__1::vector<std::__1::thread, std::__1::allocator<std::__1::thread> >::emplace_back<void (CScheduler::*)(), std::__1::reference_wrapper<CScheduler> >' requested here
        microThreads.emplace_back(&CScheduler::serviceQueue, std::ref(microTasks));
                     ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/type_traits:1043:5: note: '~__nat' has been explicitly marked deleted here
    ~__nat() = delete;
    ^
```

**Comment by paveljanik** [comment]
Compile error fixed here.

**Comment by sipa** [comment]
Needs rebase.

**Comment by theuni** [comment]
rebased and squashed in the compile fix.

**Comment by laanwj** [comment]
>  Nearly every thread uses separate interrupt/shutdown calls for exactly that reason.

There's another reason, and that is that interrupt is supposed to be an instant operation at the caller site, and shutdown is allowed to wait. This allows parallelism in the shutdown process:

- Send module A interrupt  (A starts terminating its threads, which may take some time)
- Send module B interrupt  (B starts terminating its threads, which may take some time)
...
- Module A shutdown (waits for A's threads to join)
- Module B shutdown (waits for B's threads to join)

Sure, the naming may not be optimal, but I think the approach makes sense.

**Comment by laanwj** [comment]
Travis does keep failing. There may be an issue after all.

**Comment by theuni** [comment]
Hmm, travis was happy before the rebase. I'll try to reproduce locally, maybe something changed.

**Comment by theuni** [comment]
The travis failures were caused by the rebase after all. ab48c5e72156b34300db4a6521cb3c9969be3937 introduced a new test that relied on the interruptible semantics. I've added a commit that should fix that up.

**Comment by btcdrak** [comment]
needs rebase again :( 
we should really get this merged

File: src/init.cpp [hunk 1]
```diff
@@ -1078,7 +1086,7 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
 
     // Start the lightweight task scheduler thread
     CScheduler::Function serviceLoop = boost::bind(&CScheduler::serviceQueue, &scheduler);
-    threadGroup.create_thread(boost::bind(&TraceThread<CScheduler::Function>, "scheduler", serviceLoop));
+    scheduler_thread = std::thread(std::bind(&TraceThread<CScheduler::Function>, "scheduler", serviceLoop));
```
Comment (JeremyRubin) [inline src/init.cpp]: Nit:

I find the semantics of bind to be pretty confusing confusing. If we're cleaning up threading semantics I think it might be clearer to do as follows:

scheduler_thread = std::thread([](){ TraceThread("scheduler", CScheduler::serviceQueue(scheduler)); };

This should also allow you to get rid of the explicit template param.

File: src/net.cpp [hunk 2]
```diff
@@ -120,6 +120,12 @@ boost::condition_variable messageHandlerCondition;
 static CNodeSignals g_signals;
 CNodeSignals& GetNodeSignals() { return g_signals; }
 
+static std::condition_variable net_interrupt_cond;
+static std::mutex cs_net_interrupt;
+static std::atomic<bool> net_interrupted(false);
```
Comment (JeremyRubin) [inline src/net.cpp]: This is OK -- initially, I was confused as to why you use an atomic var for the condition variable, as usually access is guarded by the lock so atomic would be overkill. But in this case, it's fine.

May be worth writing a profile at some point for if it is better to use a cond+atomic or to lock on writing the condition externally, or just throwing a comment in here as to why you're using an atomic.

File: src/net.cpp [hunk 3]
```diff
@@ -2176,16 +2180,16 @@ void StartNode(boost::thread_group& threadGroup, CScheduler& scheduler)
...
-    threadGroup.create_thread(boost::bind(&TraceThread<void (*)()>, "addcon", &ThreadOpenAddedConnections));
+    open_added_connections_thread = std::thread(std::bind(&TraceThread<void (*)()>, "addcon", &ThreadOpenAddedConnections));
 
     // Initiate outbound connections
-    threadGroup.create_thread(boost::bind(&TraceThread<void (*)()>, "opencon", &ThreadOpenConnections));
+    open_connections_thread = std::thread(std::bind(&TraceThread<void (*)()>, "opencon", &ThreadOpenConnections));
 
     // Process messages
-    threadGroup.create_thread(boost::bind(&TraceThread<void (*)()>, "msghand", &ThreadMessageHandler));
+    message_handler_thread = std::thread(std::bind(&TraceThread<void (*)()>, "msghand", &ThreadMessageHandler));
```
Comment (JeremyRubin) [inline src/net.cpp]: repeating comment about using lambdas.

File: src/checkqueue.h [hunk 4]
```diff
@@ -145,12 +154,14 @@ class CCheckQueue
     //! Add a batch of checks to the queue
     void Add(std::vector<T>& vChecks)
     {
-        boost::unique_lock<boost::mutex> lock(mutex);
-        BOOST_FOREACH (T& check, vChecks) {
-            queue.push_back(T());
-            check.swap(queue.back());
+        {
```
Comment (JeremyRubin) [inline src/checkqueue.h]: Just noting that there is a slight functional change here, the lock is released earlier but this seems a-OK.

File: src/checkqueue.h [hunk 5]
```diff
@@ -9,9 +9,8 @@
 #include <vector>
 
 #include <boost/foreach.hpp>
-#include <boost/thread/condition_variable.hpp>
-#include <boost/thread/locks.hpp>
-#include <boost/thread/mutex.hpp>
+#include <condition_variable>
```
Comment (JeremyRubin) [inline src/checkqueue.h]: It might be worth holding off on this commit given that my Lockfree Checkqueue PR also touches all these places; assuming i can get that merged, and it also swaps in std::thread.

File: src/init.cpp [hunk 6]
```diff
@@ -172,6 +172,10 @@ static std::thread import_thread;
 static std::thread flush_wallet_thread;
 #endif
 
+static bool fHaveGenesis = false;
+static boost::mutex cs_GenesisWait;
+static CConditionVariable condvar_GenesisWait;
```
Comment (JeremyRubin) [inline src/init.cpp]: CConditionVariable is a boost::condition_variable!!

File: src/net.cpp [hunk 7]
```diff
@@ -1966,7 +1966,7 @@ void ThreadMessageHandler()
         }
 
         if (fSleep)
-            messageHandlerCondition.timed_wait(lock, boost::posix_time::microsec_clock::universal_time() + boost::posix_time::milliseconds(100));
+            messageHandlerCondition.wait_for(lock, std::chrono::milliseconds(100), []()->bool {return net_interrupted; });
```
Comment (JeremyRubin) [inline src/net.cpp]: nit: there is no interruption point in the new code

File: src/net.cpp [hunk 8]
```diff
@@ -1910,8 +1910,8 @@ bool OpenNetworkConnection(const CAddress& addrConnect, bool fCountFailure, CSem
 
 void ThreadMessageHandler()
 {
-    boost::mutex condition_mutex;
-    boost::unique_lock<boost::mutex> lock(condition_mutex);
+    std::mutex condition_mutex;
```
Comment (JeremyRubin) [inline src/net.cpp]: Do these do anything? Can you just kill them? Was that mutex supposed to be (originally) static?

File: src/net.cpp [hunk 9]
```diff
@@ -1115,7 +1115,7 @@ static void AcceptConnection(const ListenSocket& hListenSocket) {
 void ThreadSocketHandler()
 {
     unsigned int nPrevNodeCount = 0;
-    while (true)
+    while (!net_interrupted)
```
Comment (JeremyRubin) [inline src/net.cpp]: This is a good commit; throwing generates bad code 👍

File: src/net.cpp [hunk 10]
```diff
@@ -130,6 +130,16 @@ std::thread open_added_connections_thread;
...
 std::thread message_handler_thread;
 
+static void InterruptibleSleep(uint64_t n)
+{
+    bool ret = false;
+    {
+        std::unique_lock<std::mutex> lock(cs_net_interrupt);
+        ret = net_interrupt_cond.wait_for(lock, std::chrono::milliseconds(n), []()->bool{return net_interrupted; });
+    }
+    interruption_point(ret);
```
Comment (JeremyRubin) [inline src/net.cpp]: might be better (although less clean) to use a noexcept version that returns a bool.

File: src/utiltime.cpp [hunk 11]
```diff
@@ -56,20 +59,7 @@ int64_t GetLogTimeMicros()
...
- */
-#if defined(HAVE_WORKING_BOOST_SLEEP_FOR)
-    boost::this_thread::sleep_for(boost::chrono::milliseconds(n));
-#elif defined(HAVE_WORKING_BOOST_SLEEP)
-    boost::this_thread::sleep(boost::posix_time::milliseconds(n));
-#else
-//should never get here
-#error missing boost sleep implementation
-#endif
+    std::this_thread::sleep_for(std::chrono::milliseconds(n));
```
Comment (JeremyRubin) [inline src/utiltime.cpp]: 👍

File: src/test/transaction_tests.cpp [hunk 12]
```diff
@@ -470,12 +470,14 @@ BOOST_AUTO_TEST_CASE(test_big_witness_transaction) {
 
     // check all inputs concurrently, with the cache
     PrecomputedTransactionData txdata(tx);
-    boost::thread_group threadGroup;
+    std::vector<std::thread> threadGroup;
```
Comment (JeremyRubin) [inline src/test/transaction_tests.cpp]: same comment; this is fixed in my WIP lockfree stuff so may want to leave it assuming I get mine merged.

**Comment by JeremyRubin** [comment]
@theuni I've gone through and I'll give you a "tentative" utack. Mostly looks fine.

I'd suggest mostly:

- checking typedefs which secretly use boost::*
- Getting rid of the exception throwing interruption point. It's a bad default.
- Getting rid of std::bind in favor of lambdas.
- depending on status of my wip scriptcheck queue stuff, leaving those two commits out.

File: src/init.cpp (hunk 1 above)
Comment (theuni) [inline src/init.cpp]: I don't think it's possible to do this without a bind, as TraceThread requires a bound instance, though I'd be happy to be proven wrong :)

File: src/net.cpp (hunk 2 above)
Comment (theuni) [inline src/net.cpp]: Right, that's not at all clear from the individual commits. The atomic is used as the break condition for a few condvars, as well as the loop break condition in a few places. Rather than using separate bools for each, I just lumped them all into this one.

Will add a comment.

File: src/net.cpp (hunk 3 above)
Comment (theuni) [inline src/net.cpp]: Yep, these don't need the binds. Will fix.

File: src/net.cpp (hunk 8 above)
Comment (theuni) [inline src/net.cpp]: The condvar has to use some lock, I assume this was just done to avoid creating a mutex/condvar per loop iteration. The mutex doesn't need to guard anything since the wake condition is atomic.

Sure, they could be static.

File: src/net.cpp (hunk 10 above)
Comment (theuni) [inline src/net.cpp]: I agree, these were only kept as-is to avoid scope creep.

**Comment by theuni** [comment]
@JeremyRubin Thanks for the review! To your points:
- I agree somewhat on the typedefs, but I'd like to avoid this PR getting bigger and bigger. So let's do these case-by-case.
- I don't think we can get rid of the exception throwing yet. I hate the interruption_point logic, but I don't think it's realistic to remove it all in one PR. I certainly agree with minimizing it though.
- Agree on losing the binds where possible
- Without the scriptcheck changes, this can't be tested as we can't shutdown. So I'm afraid we'll have to step on eachother's toes until one is merged. Obviously I'm happy to drop the scriptcheck changes and adapt to the new stuff if it goes in first.

File: src/init.cpp (hunk 1 above)
Comment (JeremyRubin) [inline src/init.cpp]: Github weirdly formatted my comment so you couldn't see the lambda.

    scheduler_thread = std::thread([&]{
        TraceThread("scheduler", [&]{scheduler.serviceQueue();} );
     };

Does something like that not work? Maybe it's not much better than bind...

Comment (theuni) [inline src/init.cpp]: Ah, I didn't realize that you meant to replace the binds with lambdas. I'm not sure that I find this any less confusing, but I don't really mind either way:
```c++
scheduler_thread = std::thread([&scheduler] {
    TraceThread("scheduler", [&scheduler] {
        scheduler.serviceQueue();
    });
});
```

**Comment by JeremyRubin** [comment]
@theuni The only one I think you need to fix for this pr is `CConditionVariable`. Were there others you're worried about?

No worries about the scriptcheck stuff, we'll cross that bridge when it comes.

File: src/init.cpp (hunk 1 above)
Comment (theuni) [inline src/init.cpp]: Heh, didn't see your response before I posted, but looks like we're on the same page. Either one is fine by me.

**Comment by theuni** [comment]
@JeremyRubin Fixing up CConditionVariable means doing lots of mutex/lock replacements all over the place. I'd really rather do that as a follow-up to keep this one minimal.

Now that we're not using any boost::thread's, I don't believe there's any major functional difference.

**Comment by JeremyRubin** [comment]
@theuni sounds fine, this one is nuking boost::thread not boost::condition_variable after all :)

File: src/net.cpp (hunk 8 above)
Comment (JeremyRubin) [inline src/net.cpp]: So I've actually discovered a nice little trick for this -- 


    struct dummy_mutex {
    void lock(){};
    void unlock(){};
     };

if you want to make it more clear that it's a worthless lock.

**Comment by btcdrak** [comment]
needs rebase

File: src/bitcoin-cli.cpp [hunk 13]
```diff
@@ -294,6 +294,8 @@ int CommandLineRPC(int argc, char *argv[])
     catch (const boost::thread_interrupted&) {
```
Comment (laanwj) [inline src/bitcoin-cli.cpp]: Does this still need to catch `boost::thread_interrupted`? (another one below)

Comment (theuni) [inline src/bitcoin-cli.cpp]: No, see "future changes" in the OP. There are a bunch of these to get rid of. I guess it doesn't make sense to do that later, I'll add it here.

Comment (laanwj) [inline src/bitcoin-cli.cpp]: Sorry, yes you mention it there. It just looked strange to me to add something instead of replace.

**Comment by fanquake** [comment]
Concept ACK. Needs a rebase. Tagging for 0.14.0

**Comment by fanquake** [comment]
There also some comments/examples about using boost thread [in scheduler](https://github.com/bitcoin/bitcoin/blob/master/src/scheduler.h#L9) that could be nuked.

**Comment by theuni** [comment]
Closing for now, need to get a specific chunk of this in first.
